source/ingress.go

```diff
@@ -63,12 +66,27 @@ type ingressSource struct {
 }
 
 // NewIngressSource creates a new ingressSource with the given config.
-func NewIngressSource(ctx context.Context, kubeClient kubernetes.Interface, namespace, annotationFilter string, fqdnTemplate string, combineFqdnAnnotation bool, ignoreHostnameAnnotation bool, ignoreIngressTLSSpec bool, ignoreIngressRulesSpec bool, labelSelector labels.Selector) (Source, error) {
+func NewIngressSource(ctx context.Context, kubeClient kubernetes.Interface, namespace, annotationFilter string, fqdnTemplate string, combineFqdnAnnotation bool, ignoreHostnameAnnotation bool, ignoreIngressTLSSpec bool, ignoreIngressRulesSpec bool, labelSelector labels.Selector, ingressClassNames []string) (Source, error) {
 	tmpl, err := parseTemplate(fqdnTemplate)
 	if err != nil {
 		return nil, err
 	}
 
+	// ensure that ingress class is only set in either the ingressClassNames or
+	// annotationFilter but not both
+	if ingressClassNames != nil && annotationFilter != "" {
+		selector, err := getLabelSelector(annotationFilter)
+		if err != nil {
+			return nil, err
+		}
+
+		requirements, _ := selector.Requirements()
+		for _, requirement := range requirements {
+			if requirement.Key() == "kubernetes.io/ingress.class" {
```

Comment:
This might cause an issues I think 🤔 
Let's assume that you have two Ingress controllers one internal and one external, and that the internal is started with `--ingress-class` but the external one should listen on annotations, wouldn't this then break???

Reply:
Isn't the annotation deprecated? 
If so, so should the support for it.
You can't deploy an ingress resource with both annotation and ingressclassname but I guess you can run two ingresses.
I think one of my AKS labs has such setup

Reply:
@mateuszdrab 
> **You can't deploy an ingress resource with both annotation and ingressclassname** but I guess you can run two ingresses.

You can. Kubernetes 1.19 complains when you try to deploy both at the same time, but 1.20+ supports it.

Reply:
> @mateuszdrab 
> 
> > **You can't deploy an ingress resource with both annotation and ingressclassname** but I guess you can run two ingresses.
> 
> 
> 
> You can. Kubernetes 1.19 complains when you try to deploy both at the same time, but 1.20+ supports it.
> 
> 

Oh, ok. Didn't realize - I guess whatever is the official statement about supporting it from the Kubernetes team is what should be followed here in the code implementation.

Reply:
Also not all ingress controllers have gotten with the times and support ingressClassName, some still operate on the annotation (Kong for example).  And deprecated != deleted.

Reply:
I think the normal case is to fallback to the annotation if the class is not set. Please change the behavior to fallback

Reply:
`filterByIngressClass()` does the fallback logic when considering whether an `Ingress` matches the requirement set by the `--ingress-class` flag. For that reason, passing both an `--ingress-class` flag and a `kubernetes.io/ingress.class` annotation filter doesn't make sense: it would exclude those ingresses that don't match the annotation filter. Thus, this check is a reasonable validation against the admin passing nonsensical flag combinations.